### PR TITLE
Aggro Sync

### PR DESCRIPTION
**Pull request by Protryon**
This PR massively increases network-bound sync performance and significantly reduces any chances of forks not being considered.

The old sync process:
1. Request sync hashes from peers based on block locator hashes
2. Randomly request each novel sync block from a random peer who has it
3. Repeat

The new sync process:
1. Request sync hashes from peers based on block locator hashes
2. For each peer that responds, broadcast back their highest sync hash, request all blocks we don't have at that instant
3. Repeat step # 2 until no one responds for a while

This has the following effects:
* Continuous sync process eliminates cyclical performance observed/expected in old sync process.
* Increase in the number of received duplicate blocks (from effectively zero to scaling linearly by number of peers)
* An effective, naive negotiation for any peers on really distant forks to still be able to send blocks.
* Pulling forks that are ultimately suboptimal, in their entirety.

My current node synced to canon tip from origin in ~18 hours of runtime -- unclean test. Has ~40K forked blocks from the process (not forks, blocks not in canon).

### CODE REVIEW

**Comment by ljedrz** [comment]
I've started testing this branch and I can immediately say we should probably downgrade some of the exsiting `WARN` logs to `TRACE` now, especially the following 2:
```
WARN received unexpected or late sync block from X
WARN failed receiving block: PreExistingBlock
```

File: network/src/inbound/cache.rs [hunk 1]
```diff
@@ -14,27 +14,30 @@
...
             queue: CircularQueue::with_capacity(N),
         }
     }
 }
 
-impl<const N: usize> Cache<N> {
+impl<const N: usize, const S: usize> Cache<N, S> {
     fn hash_block(payload: &[u8]) -> u64 {
-        hash64(&payload[..BlockHeader::size()])
+        hash64(&payload[..S.min(payload.len())])
```
Comment (ljedrz) [inline network/src/inbound/cache.rs:40]: is it possible for a `Block` to be smaller than the header :thinking:?

File: network/src/sync/blocks/batched.rs [hunk 2]
```diff
@@ -42,86 +35,20 @@ struct SyncBlock {
...
-            .await?
-            .into_iter()
-            .map(|x| BlockHeaderHash(x.bytes::<32>().unwrap()))
-            .collect::<Vec<_>>();
+        let block_locator_hashes = SyncBase::block_locator_hashes(&self.base.node).await?;
         let mut future_set = vec![];
 
         for peer in sync_nodes.iter() {
-            if let Some(handle) = self.node.peer_book.get_peer_handle(peer.address) {
+            if let Some(handle) = self.base.node.peer_book.get_peer_handle(peer.address) {
```
Comment (ljedrz) [inline network/src/sync/blocks/batched.rs:51]: this is not strictly related to this PR, but: since we only need the peers' addresses, it is probably preferable to put only those in the `sync_nodes` vector, to reduce the amount of memory and atomic operations

File: network/src/sync/blocks/aggro.rs [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+                .collect()
+        };
+
+        let hash_requests_sent = self.send_sync_messages(peer_syncs).await?;
+
+        if hash_requests_sent == 0 {
+            return Ok(());
+        }
+
+        let received_hashes = Arc::new(RwLock::new(Cache::<1024, 32>::default()));
```
Comment (ljedrz) [inline network/src/sync/blocks/aggro.rs:89]:
```suggestion
        let received_hashes = Arc::new(RwLock::new(Cache::<1024, 32>::default()));
```
it would be best to refer to `BLOCK_HEADER_SIZE` instead of `32` here

File: network/src/sync/blocks/aggro.rs [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+
+                            let early_block_states = match node.storage.get_block_states(&hashes_trimmed[..]).await {
+                                Ok(x) => x,
+                                Err(e) => {
+                                    warn!("failed to get block states: {:?}", e);
+                                    return;
+                                }
+                            };
+
+                            let blocks: IndexSet<_> = {
```
Comment (ljedrz) [inline network/src/sync/blocks/aggro.rs:137]: do we need an `IndexSet` (as opposed to a plain `HashSet`) here?

File: network/src/sync/blocks/base.rs [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+    pub async fn find_sync_nodes(&self) -> Result<Vec<Peer>> {
+        let our_block_height = self.node.storage.canon().await?.block_height;
+        let mut interesting_peers = vec![];
+        for mut node in self.node.peer_book.connected_peers_snapshot().await {
+            let judge_bad = node.judge_bad();
+            if !judge_bad && node.quality.block_height as usize > our_block_height + 1 {
+                interesting_peers.push(node);
+            }
+        }
+        interesting_peers.sort_by(|x, y| y.quality.block_height.cmp(&x.quality.block_height));
```
Comment (ljedrz) [inline network/src/sync/blocks/base.rs:51]:
```suggestion
        interesting_peers.sort_by_key(|x| std::cmp::Reverse(x.quality.block_height));
```
`clippy` will suggest this

File: network/src/sync/blocks/batched.rs [hunk 6]
```diff
@@ -133,55 +60,25 @@ impl SyncMaster {
...
+                    SyncInbound::Block(..) => {
+                        warn!("received sync block prematurely");
+                    }
                 }
-            }
-            //todo: fail if peer sends > 1 block hash packet
-            received_block_hashes.len() >= max_message_count
-        })
-        .await;
+                //todo: fail if peer sends > 1 block hash packet
```
Comment (ljedrz) [inline network/src/sync/blocks/batched.rs:78]: is this `todo` still applicable?

File: network/src/sync/blocks/aggro.rs [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+use anyhow::*;
+use indexmap::IndexSet;
+use snarkos_storage::Digest;
+use snarkvm_dpc::BlockHeaderHash;
+use tokio::{
+    sync::{mpsc, RwLock},
+    time::Instant,
+};
+
+pub struct SyncAggro {
```
Comment (ljedrz) [inline network/src/sync/blocks/aggro.rs:29]: having seen all the `SyncX` objects I know what their names mean, but it would be nice to have short doc comments on them for future readers

**Review by ljedrz — COMMENTED** [review]
Left some questions and suggestions; I'll run some more tests (the initial runs were very promising) in the meantime.

File: network/src/inbound/cache.rs (hunk 1 above)
Comment (Protryon) [reply network/src/inbound/cache.rs:40]: no, but we use digests now -- and truncation is preferable to panicing in general

File: network/src/sync/blocks/batched.rs (hunk 2 above)
Comment (Protryon) [reply network/src/sync/blocks/batched.rs:51]: yeah, not worth doing since we may need more in the future, and the costs here are trivial

File: network/src/sync/blocks/aggro.rs (hunk 4 above)
Comment (Protryon) [reply network/src/sync/blocks/aggro.rs:137]: no -- but i appreciate the added determinism as a hedge against non-determinism bugs -- and as usual, not perf critical

File: network/src/sync/blocks/batched.rs (hunk 6 above)
Comment (Protryon) [reply network/src/sync/blocks/batched.rs:78]: it was always just an idea, and remains a valid idea

File: network/src/sync/blocks/aggro.rs (hunk 3 above)
Comment (Protryon) [reply network/src/sync/blocks/aggro.rs:89]: it's block hash size, not header size. don't think we have a constant for that atm, just default digest size.

File: network/src/sync/blocks/base.rs (hunk 5 above)
Comment (Protryon) [reply network/src/sync/blocks/base.rs:51]: `clippy` didn't suggest, avoiding for now.

File: network/src/sync/blocks/aggro.rs [hunk 8]
```diff
@@ -0,0 +1,202 @@
...
+                        });
+                    }
+                }
+                false
+            })
+            .await;
+
+        let sync_addresses = sync_nodes.iter().map(|x| x.address).collect::<Vec<_>>();
+
+        self.base.cancel_outstanding_syncs(&sync_addresses[..]).await;
```
Comment (niklaslong) [inline network/src/sync/blocks/aggro.rs:192]: It wasn't immediately obvious to me why this needs to be called at the end of every sync, it would be nice to comment on this here? (It may also not be necessary if state is tracked on the sync layer.)

**Review by niklaslong — APPROVED** [review]
This is working really nicely for me locally, left a few comments 🚀

File: network/src/peers/peer/outbound_handler.rs [hunk 9]
```diff
@@ -208,13 +208,13 @@ impl Peer {
...
-                    warn!("received unexpected or late sync block from {}", self.address);
+                    trace!("received unexpected or late sync block from {}", self.address);
                 }
                 Ok(PeerResponse::None)
             }
             PeerAction::ExpectingSyncBlocks(amount) => {
-                self.quality.remaining_sync_blocks = amount;
-                self.quality.total_sync_blocks = amount;
+                self.quality.remaining_sync_blocks += amount;
+                self.quality.total_sync_blocks += amount;
```
[the remark this answers is not in the document]
Comment (Protryon) [reply network/src/peers/peer/outbound_handler.rs:217]: yes, but in a future pr

File: network/src/sync/blocks/aggro.rs (hunk 8 above)
Comment (Protryon) [reply network/src/sync/blocks/aggro.rs:192]: it notifies each peer to reset tracking state. Agreed.
